parley/src/layout/editor.rs

```diff
@@ -485,38 +614,210 @@ where
     }
+
+    /// Update the layout unconditionally.
+    fn update_layout(&mut self) {
+        self.editor.update_layout(self.font_cx, self.layout_cx);
+    }
 }
 
 impl<T> PlainEditor<T>
 where
     T: Brush + Clone + Debug + PartialEq + Default,
 {
-    /// Run a series of [`PlainEditorTxn`] methods, updating the layout
-    /// if necessary.
-    pub fn transact(
+    // --- MARK: PlainEditor API ---
+    /// Run a series of [`PlainEditorDriver`] methods.
+    ///
+    /// This is a utility shorthand around [`driver`](Self::driver); it
+    /// does not have any different behaviour.
+    pub fn drive<R>(
         &mut self,
         font_cx: &mut FontContext,
         layout_cx: &mut LayoutContext<T>,
-        callback: impl FnOnce(&mut PlainEditorTxn<'_, T>),
-    ) {
-        let mut txn = PlainEditorTxn {
+        callback: impl FnOnce(&mut PlainEditorDriver<'_, T>) -> R,
+    ) -> R {
+        let mut drv = self.driver(font_cx, layout_cx);
+        callback(&mut drv)
+    }
+
+    /// Run a series of [`PlainEditorDriver`] methods.
+    ///
+    /// This type is only used to simplify methods which require both
+    /// the editor and the provided contexts.
+    pub fn driver<'drv>(
+        &'drv mut self,
+        font_cx: &'drv mut FontContext,
+        layout_cx: &'drv mut LayoutContext<T>,
+    ) -> PlainEditorDriver<'drv, T> {
+        PlainEditorDriver {
             editor: self,
             font_cx,
             layout_cx,
-        };
-        callback(&mut txn);
-        txn.refresh_layout();
+        }
+    }
+
+    /// If the current selection is not collapsed, returns the text content of
+    /// that selection.
+    pub fn selected_text(&self) -> Option<&str> {
+        if !self.selection.is_collapsed() {
+            self.text().get(self.selection.text_range())
+        } else {
+            None
+        }
+    }
+
+    /// Get rectangles representing the selected portions of text.
+    pub fn selection_geometry(&self) -> Vec<Rect> {
+        self.selection.geometry(&self.layout)
     }
 
-    /// Make a cursor at a given byte index
+    /// Get a rectangle representing the current caret cursor position.
+    pub fn cursor_geometry(&self, size: f32) -> Option<Rect> {
+        Some(self.selection.focus().geometry(&self.layout, size))
+    }
+
+    /// Borrow the text content of the buffer.
+    pub fn text(&self) -> &str {
+        &self.buffer
+    }
+
+    /// Get the current `Generation` of the layout, to decide whether to draw.
+    ///
+    /// You should store the generation the editor was at when you last drew it, and then redraw
+    /// when the generation is different (`Generation` is [`PartialEq`], so supports the equality `==` operation).
+    pub fn generation(&self) -> Generation {
+        self.generation
+    }
+
+    /// Replace the whole text buffer.
+    pub fn set_text(&mut self, is: &str) {
+        self.buffer.clear();
+        self.buffer.push_str(is);
+        self.layout_dirty = true;
+    }
```

Comment:
We may want to add


```rust
debug_assert!(!self.is_composing());
```

here too. This was missed in [xilem#762](https://github.com/linebender/xilem/pull/762).

Reply:
Agreed, will do. This gets a bit messy for Xilem; this can happen based on external factors to IME, so I'm not certain how to resolve that...

Reply:
Good point. We already warn in, e.g., `TextArea` that doing that is likely disruptive if the user is currently focused on the widget. We may be able to just clear compose, set text, and set the compose again, though the compose would likely end up in a somewhat strange location. Alternatively, a more rigorous clear compose, end IME, set text, start IME.

We can discuss this further outside of Parley.

Reply:
Agreed, I don't think this needs to be blocking here.